frontend/ui/lib/widgets/retryable_async_loadable.dart

```diff
@@ -0,0 +1,88 @@
+import 'package:flutter/material.dart';
+
+/// Provides loading and retrying functionality.
+/// This should only be used with an idempotent async function.
+class RetryableAsyncLoadable<T> extends StatefulWidget {
+  final _OnLoad<T> onLoad;
+  final _LoadFunction<T> loadFunction;
+  final String errorMessage;
+
+  RetryableAsyncLoadable({
+    key,
+    @required this.onLoad,
+    @required this.loadFunction,
+    @required this.errorMessage,
+  }) : super(key: key);
+
+  @override
+  State createState() =>
+      _RetryableAsyncLoadableState(onLoad, loadFunction, errorMessage);
+}
+
+typedef _LoadFunction<T> = Future<T> Function();
+typedef _OnLoad<T> = Widget Function(T);
+
+class _RetryableAsyncLoadableState<T> extends State<RetryableAsyncLoadable<T>> {
+  final _OnLoad<T> onLoad;
+  final _LoadFunction<T> loadFunction;
+  final String errorMessage;
+  bool loading = true;
+  Exception error;
+  T t;
+
+  _RetryableAsyncLoadableState(
+      this.onLoad, this.loadFunction, this.errorMessage) {
+    preformAction();
```

Comment:
Are you allowed to call an async function from the constructor? (I know next to nothing about Dart.)

Reply:
yeah, you can call an async function whenever, you just cant `await` an async function unless you're inside an async function. Dart itself is single threaded, so starting an async function like this just schedules the action on the event loop.